File: 0070-climbing-stairs.py
"""
70. Climbing Stairs

Submitted: January 9th, 2025

Runtime: 0 ms (beats 100.00%)
Memory: 17.87 MB (beats 30.40%)
"""

class Solution:
    def climbStairs(self, n: int) -> int:
        a = 0
        b = 1
        for _ in range(n):
            c = a + b
            a = b
            b = c
        return b

File: 0070-climbing-stairs.cpp
/*
70. Climbing Stairs

Submitted: January 9th, 2025

Runtime: 0 ms (beats 100.00%)
Memory: 7.97 MB (beats 43.48%)
*/

class Solution {
public:
    int cache[45] = { 0 };
    int climbStairs(int n) {
        if (n == 0 || n == 1) {
            return 1;
        }
        if (cache[n - 1] != 0) return cache[n - 1];
        int res = climbStairs(n - 2) + climbStairs(n - 1);
        if (cache[n - 1] == 0) cache[n - 1] = res;
        return res;
    }
};